### ISSUE TITLE
find a better way to display GitHub links E597

### ISSUE THREAD

**Issue by only1chunts** [gigascience/gigadb-website#1841]
## User story

>As a website user  
>I want to see the GitHub repository link & its archival version link together
>So that its obvious they are tightly related

NB - current the SWH link is added as "additional information" and the GitHub link is added as "GitHub archive" and they are listed separately on the website.

## Acceptance criteria

>Given I am on a GigaDB dataset page with a link to a GitHub repository  
>When I examine the GitHub link 
>Then I am shown two options; (1) a link to the current GitHub repository, and (2) a link to the archival version of the repository hosted at Software Heritage Library (SWH).

## Additional Info
This will likely require changes to how we add the information to the database. I would like to propose we expand the "External links" table (I think its called External_Link in the schema) to include an optional secondary link called something like "archival version link" or similar.

This should be considered part of Epic #597 

## Product Backlog Item Ready Checklist

* [ ] Business value is clearly articulated
* [ ] Item is understood enough by the IT team so it can make an informed decision as to whether it can complete this item
* [ ] Dependencies are identified and no external dependencies would block this item from being completed
* [ ] At the time of the scheduled sprint, the IT team has the appropriate composition to complete this item
* [ ] This item is estimated and small enough to comfortably be completed in one sprint
* [ ] Acceptance criteria are clear and testable
* [ ] Performance criteria, if any, are defined and testable
* [ ] The Scrum team understands how to demonstrate this item at the sprint review

## Product Backlog Item Done Checklist

* [ ] Item(s) in increment pass all Acceptance Criteria
* [ ] Code is refactored to best practices and coding standards
* [ ] Documentation is updated as needed
* [ ] Data security has not been compromised (with particular reference to the personal information we hold in GigaDB)
* [ ] No deviation from the team technology stack and software architecture has been introduced
* [ ] The product is in a releasable state (i.e. the increment has not broken anything)

**Comment by alli83** [comment]
@only1chunts would it be possible to have more information?

**Comment by only1chunts** [comment]
Context: If an author has created a tool and stored the code for that in GitHub, then we want to capture a link to that "live" GitHub repository, however as we're publishing at a particular point in time so we need to also capture a "version-of-record" or snapshot of that repository at the time of publication because the "live" github is liable to change over time after publication. Software Heritage Library offer that service, so we want to be able to hold both the "live" GitHub link AND the "version-of-record" Software Heritage Archive link (SWHID). Having them as two separate entities in the external_link table does work and that is the bare minimum for us, so what you have done/are doing in the other tickets (#299) to make that possible is sufficient.... But IF there is a way to be able to make the connection between those two separate external_link objects to better demonstrate that the SWHID is infact a snapshot of the GitHub repo, then that would make for a better user experience.

I dont know how to do this in the back-end, but from a user perspective on the front end it would be fab to have something like this (NB the red outline is just to draw your attention to the relevant part it wouldn't be visible on the real page!)

![Image](https://github.com/user-attachments/assets/9cfd8cd3-ac13-4d9f-875c-cfb22275f7e1)

**Comment by alli83** [comment]
@only1chunts thanks! I'll look into it.

**Comment by alli83** [comment]
@only1chunts  when you say SWHID, do you mean Software Heritage Archive (SWHA)? because we don't have SWHID as external link type

**Comment by only1chunts** [comment]
yes, sorry a SWHID is a SoftWare Heritage Archive IDentifier, so SWHA is the provider of a SWHID :-D

**Comment by alli83** [comment]
@only1chunts is it only for "Authors code repositories" or the "cited" ones too?

**Comment by only1chunts** [comment]
I think it makes sense to do it for both.

**Comment by alli83** [comment]
@only1chunts will the link always contain "origin_url" in the url? or do I need to scrape the html page?

**Comment by alli83** [comment]
@rija @only1chunts If I try to retrieve the origin URL from a snapshot, it could be pretty heavy because I'd need to list all the branches, then for each branch, fetch the revision and check if the metadata contains the origin

If I have a revision, I can directly check if the metadata (form the software archive API) contains the origin but about the snapshot it's a bit tricky

**Comment by alli83** [comment]
@only1chunts @rija Could you please confirm that it will be when it is selected: Software Heritage Archive for type?
Also, what is the format of the URL:  https://archive.softwareheritage.org/swh:1:dir:...?

**Comment by only1chunts** [comment]
@alli83 I'm not entirely sure I am following what you are trying to do with fetching things? I dont know what needs to be fetched from anywhere, we (curators) provide the SWHID as we do for all the other external links. All we need is the place in the database to store them and the website to display them.
I am slightly torn between locking us into using SWH, or being able to provide other archive links (its also possible for authors to provide the snapshot of their repository in Zenodo or other repositories).
Is it possible to have it generic enough to allow any URL? but still display it in the way shown in the mockup screenshot above?

Also, just as an FYI - bioregistry.io do include [SWH as a known prefix](https://bioregistry.io/registry/swh)

**Comment by alli83** [comment]
@only1chunts on external link form: Sometimes the SWHID will have the origin in the URL, so in that case it'll be easy to extract and use it to look up the external link with that origin ID and link it. But sometimes the origin isn’t present in the SWHID URL. In that case, if it’s a revision or snapshot type, we’ll need to try using the API to retrieve the origin URL. If you want, we can talk more about it during the next call

example of SWHID:
https://archive.softwareheritage.org/swh:1:dir:4b825dc642cb6eb9a060e54bf8d69288fbee4904;origin=https://github.com/example
https://archive.softwareheritage.org/swh:1:rev:abcdef1234567890/
https://archive.softwareheritage.org/swh:1:snp:1234567890abcdef/

**Comment by only1chunts** [comment]
The curators will be responsible for adding the full URL, including the
origin. The reason there are missing ones at the moment is due to the web
interface preventing long URLs being entered via the admin pages. (The
upload tool pushes them in fine, hence there are some in there).

**Comment by alli83** [comment]
@only1chunts ok thanks for the feedback. I'm working on this issue: 
I handle the case if origin is in the url => simple case
also if it's not and if it's a revision type I can use the software heritage API to check the metadata associated with it and check if origin if present.
